// powerbi-visuals-utils-chartutils/lib/label/dataLabelPointPositioner.js
/*
*  Power BI Visualizations
*
*  Copyright (c) Microsoft Corporation
*  All rights reserved.
*  MIT License
*
*  Permission is hereby granted, free of charge, to any person obtaining a copy
*  of this software and associated documentation files (the ""Software""), to deal
*  in the Software without restriction, including without limitation the rights
*  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
*  copies of the Software, and to permit persons to whom the Software is
*  furnished to do so, subject to the following conditions:
*
*  The above copyright notice and this permission notice shall be included in
*  all copies or substantial portions of the Software.
*
*  THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
*  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
*  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
*  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
*  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
*  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
*  THE SOFTWARE.
*/
export const cos45 = Math.cos(45);
export const sin45 = Math.sin(45);
export function getLabelRect(labelSize, parentPoint, position, offset) {
    switch (position) {
        case 1 /* Above */: {
            return above(labelSize, parentPoint.point, parentPoint.radius + offset);
        }
        case 2 /* Below */: {
            return below(labelSize, parentPoint.point, parentPoint.radius + offset);
        }
        case 4 /* Left */: {
            return left(labelSize, parentPoint.point, parentPoint.radius + offset);
        }
        case 8 /* Right */: {
            return right(labelSize, parentPoint.point, parentPoint.radius + offset);
        }
        case 32 /* BelowLeft */: {
            return belowLeft(labelSize, parentPoint.point, parentPoint.radius + offset);
        }
        case 16 /* BelowRight */: {
            return belowRight(labelSize, parentPoint.point, parentPoint.radius + offset);
        }
        case 128 /* AboveLeft */: {
            return aboveLeft(labelSize, parentPoint.point, parentPoint.radius + offset);
        }
        case 64 /* AboveRight */: {
            return aboveRight(labelSize, parentPoint.point, parentPoint.radius + offset);
        }
        case 256 /* Center */: {
            return center(labelSize, parentPoint.point);
        }
    }
    return null;
}
export function above(labelSize, parentPoint, offset) {
    return {
        left: parentPoint.x - (labelSize.width / 2),
        top: parentPoint.y - offset - labelSize.height,
        width: labelSize.width,
        height: labelSize.height
    };
}
export function below(labelSize, parentPoint, offset) {
    return {
        left: parentPoint.x - (labelSize.width / 2),
        top: parentPoint.y + offset,
        width: labelSize.width,
        height: labelSize.height
    };
}
export function left(labelSize, parentPoint, offset) {
    return {
        left: parentPoint.x - offset - labelSize.width,
        top: parentPoint.y - (labelSize.height / 2),
        width: labelSize.width,
        height: labelSize.height
    };
}
export function right(labelSize, parentPoint, offset) {
    return {
        left: parentPoint.x + offset,
        top: parentPoint.y - (labelSize.height / 2),
        width: labelSize.width,
        height: labelSize.height
    };
}
export function belowLeft(labelSize, parentPoint, offset) {
    return {
        left: parentPoint.x - (sin45 * offset) - labelSize.width,
        top: parentPoint.y + (cos45 * offset),
        width: labelSize.width,
        height: labelSize.height
    };
}
export function belowRight(labelSize, parentPoint, offset) {
    return {
        left: parentPoint.x + (sin45 * offset),
        top: parentPoint.y + (cos45 * offset),
        width: labelSize.width,
        height: labelSize.height
    };
}
export function aboveLeft(labelSize, parentPoint, offset) {
    return {
        left: parentPoint.x - (sin45 * offset) - labelSize.width,
        top: parentPoint.y - (cos45 * offset) - labelSize.height,
        width: labelSize.width,
        height: labelSize.height
    };
}
export function aboveRight(labelSize, parentPoint, offset) {
    return {
        left: parentPoint.x + (sin45 * offset),
        top: parentPoint.y - (cos45 * offset) - labelSize.height,
        width: labelSize.width,
        height: labelSize.height
    };
}
export function center(labelSize, parentPoint) {
    return {
        left: parentPoint.x - (labelSize.width / 2),
        top: parentPoint.y - (labelSize.height / 2),
        width: labelSize.width,
        height: labelSize.height
    };
}
export function getLabelLeaderLineEndingPoint(boundingBox, position, parentShape) {
    let x = boundingBox.left;
    let y = boundingBox.top;
    switch (position) {
        case 1 /* Above */:
            x += (boundingBox.width / 2);
            y += boundingBox.height;
            break;
        case 2 /* Below */:
            x += (boundingBox.width / 2);
            break;
        case 4 /* Left */:
            x += boundingBox.width;
            y += ((boundingBox.height * 2) / 3);
            break;
        case 8 /* Right */:
            y += ((boundingBox.height * 2) / 3);
            break;
        case 32 /* BelowLeft */:
            x += boundingBox.width;
            y += (boundingBox.height / 2);
            break;
        case 16 /* BelowRight */:
            y += (boundingBox.height / 2);
            break;
        case 128 /* AboveLeft */:
            x += boundingBox.width;
            y += boundingBox.height;
            break;
        case 64 /* AboveRight */:
            y += boundingBox.height;
            break;
    }
    return [[parentShape.point.x, parentShape.point.y], [x, y]];
}
